Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -131,6 +148,27 @@ extension SwiftLanguageServer {
       guard let self = self else { return }
       let notification = SKResponse(notification, sourcekitd: self.sourcekitd)
 
+      if notification.value?[self.keys.notification] == self.values.notification_sema_enabled {
+        self.state = .connected
+      }
+
+      if self.state == .connectionInterrupted {
+        // If we get a notification while we are restoring the connection, it means that the server has restarted.
+        // We still need to wait for semantic functionality to come back up.
+        self.state = .semanticFunctionalityDisabled
+
+        // Ask our parent to re-open all of our documents.
+        self.reopenDocuments(self)
+      }
+
+      if let error = notification.error, error.code == .connectionInterrupted {
+        self.state = .connectionInterrupted
+
+        // We don't have any open documents anymore after sourcekitd crashed.
+        // Reset the document manager to reflect that.
+        self.documentManager = DocumentManager()
```

Comment:
This mutation also needs synchronization.  It's kind of suspicious that you're doing this *after* you call reopenDocuments - couldn't that repopulate it at the wrong time?

Reply:
> This mutation also needs synchronization.

Good point. I wrapped it in `queue.async`.

> It's kind of suspicious that you're doing this after you call reopenDocuments - couldn't that repopulate it at the wrong time?

No, that ordering is correct. When `sourcekitd` crashes, `state` is `.connected`. Hence the second `if self.state == .connectionInterrupted` is skipped and the third `if let error = notification.error, error.code == .connectionInterrupted` sets `state = .connectionInterrupted`, because we receive a notification with `error.code == .connectionInterrupted`. 

When the next notification comes in, the first `if self.state == .connectionInterrupted` is taken, which then sets `state = .semanticFunctionalityDisabled`.

Finally, when semantic functionality is restored, we get a notification that satisfies the first `if notification.value?[self.keys.notification] == self.values.notification_sema_enabled`, setting `state` back to `.connected`.